bee-tangle/src/ms_tangle.rs

```diff
@@ -0,0 +1,1145 @@
+            info!("Failed to insert message {:?}", e);
+            None
+        }) {
+            let message_id = *self
+                .milestones
+                .lock()
+                .await
+                .entry(idx)
+                .or_insert(milestone)
+                .message_id();
+
+            Some(message_id)
+        } else {
+            None
+        }
+    }
+
+    /// Get the milestone from the tangle that corresponds to the given milestone index.
+    pub async fn get_milestone(&self, index: MilestoneIndex) -> Option<Milestone> {
+        self.milestones.lock().await.get(&index).cloned()
+    }
+
+    /// Get the message associated with the given milestone index from the tangle.
+    pub async fn get_milestone_message(&self, index: MilestoneIndex) -> Option<MessageRef> {
+        // TODO: use combinator instead of match
+        match self.get_milestone_message_id(index).await {
+            None => None,
+            Some(ref hash) => self.get(hash).await,
+        }
+    }
+
+    /// Get the message ID associated with the given milestone index from the tangle.
+    pub async fn get_milestone_message_id(&self, index: MilestoneIndex) -> Option<MessageId> {
+        let message_id = self.milestones.lock().await.get(&index).map(|m| *m.message_id());
+
+        // TODO: use combinator instead of match
+        match message_id {
+            Some(m) => Some(m),
+            None => Some(self.pull_milestone(index).await?),
+        }
+    }
+
+    /// Return whether the tangle contains the given milestone index.
+    pub async fn contains_milestone(&self, idx: MilestoneIndex) -> bool {
+        // Not using `||` as its first operand would keep the lock alive causing a deadlock with its second operand.
+        if self.milestones.lock().await.contains_key(&idx) {
+            return true;
+        }
+        self.pull_milestone(idx).await.is_some()
+    }
+
+    /// Get the index of the latest milestone.
+    pub fn get_latest_milestone_index(&self) -> MilestoneIndex {
+        self.latest_milestone_index.load(Ordering::Relaxed).into()
+    }
+
+    /// Update the index of the lastest milestone.
+    pub fn update_latest_milestone_index(&self, new_index: MilestoneIndex) {
+        // TODO: `fetch_max`? Swap and ensure the old is smaller?
+        self.latest_milestone_index.store(*new_index, Ordering::Relaxed);
+    }
+
+    /// Get the latest solid milestone index.
+    pub fn get_solid_milestone_index(&self) -> MilestoneIndex {
+        self.solid_milestone_index.load(Ordering::Relaxed).into()
+    }
+
+    /// Update the latest solid milestone index.
+    pub fn update_solid_milestone_index(&self, new_index: MilestoneIndex) {
+        self.solid_milestone_index.store(*new_index, Ordering::Relaxed);
+
+        // TODO: Formalise this a little better
+        let new_len = ((1000.0 + self.get_sync_threshold() as f32 * 500.0) as usize)
```

Comment:
maybe letting the cast after the product is faster. Could it happen that this overflows in that case?

Reply:
I can promise that there are far, far more expensive things to care about on this code path than an extra cast :grin: (and besides, this code only gets run intermittently anyway).